File: src/Native/ShaderReflection.cpp
#include "ShaderReflection.h"
#include "TypeReflection.h"
#include "TypeLayoutReflection.h"
#include "VariableLayoutReflection.h"
#include "VariableReflection.h"
#include "FunctionReflection.h"
#include "EntryPointReflection.h"
#include "GenericReflection.h"
#include "TypeParameterReflection.h"
#include "LayoutRules.h"
#include "ProgramCLI.h"

Native::ShaderReflection::ShaderReflection(ProgramCLI* parent, unsigned int targetIndex)
{
    if (!parent) throw std::invalid_argument("Parent cannot be null");
    
    m_parent = parent;

    // Get the native slang reflection from the program
    slang::ProgramLayout* slangProgramLayout = m_parent->GetLayout(targetIndex);

    m_native = (slang::ShaderReflection*)slangProgramLayout;
}

Native::ShaderReflection::~ShaderReflection()
{
    // m_parent is managed by ProgramCLI, so we don't delete it here
    //if (m_parent)
    //{
    //    delete m_parent;
    //    m_parent = nullptr;
    //}

    // No need to delete m_native here, as it is managed by Slang
    //if (m_native)
    //{
    //    delete m_native;
    //    m_native = nullptr;
    //}
}

Native::ProgramCLI* Native::ShaderReflection::getParent()
{
    return m_parent;
}

slang::ShaderReflection* Native::ShaderReflection::getNative()
{
    return m_native;
}

unsigned Native::ShaderReflection::getParameterCount()
{
    return m_native->getParameterCount();
}

unsigned Native::ShaderReflection::getTypeParameterCount()
{
    return m_native->getTypeParameterCount();
}

Native::TypeParameterReflection* Native::ShaderReflection::getTypeParameterByIndex(unsigned index)
{
    return new Native::TypeParameterReflection(m_native->getTypeParameterByIndex(index));
}

Native::TypeParameterReflection* Native::ShaderReflection::findTypeParameter(char const* name)
{
    return new Native::TypeParameterReflection(m_native->findTypeParameter(name));
}

Native::VariableLayoutReflection* Native::ShaderReflection::getParameterByIndex(unsigned index)
{
    return new Native::VariableLayoutReflection(m_native->getParameterByIndex(index));
}

SlangUInt Native::ShaderReflection::getEntryPointCount()
{
    return m_native->getEntryPointCount();
}

Native::EntryPointReflection* Native::ShaderReflection::getEntryPointByIndex(SlangUInt index)
{
    return new Native::EntryPointReflection(this, m_native->getEntryPointByIndex(index));
}

Native::EntryPointReflection* Native::ShaderReflection::findEntryPointByName(const char* name)
{
    return new Native::EntryPointReflection(this, m_native->findEntryPointByName(name));
}

SlangUInt Native::ShaderReflection::getGlobalConstantBufferBinding()
{
    return m_native->getGlobalConstantBufferBinding();
}

size_t Native::ShaderReflection::getGlobalConstantBufferSize()
{
    return m_native->getGlobalConstantBufferSize();
}

Native::TypeReflection* Native::ShaderReflection::findTypeByName(const char* name)
{
    return new Native::TypeReflection(m_native->findTypeByName(name));
}

Native::FunctionReflection* Native::ShaderReflection::findFunctionByName(const char* name)
{
    return new Native::FunctionReflection(m_native->findFunctionByName(name));
}

Native::FunctionReflection* Native::ShaderReflection::findFunctionByNameInType(TypeReflection* type, const char* name)
{
    return new Native::FunctionReflection(m_native->findFunctionByNameInType(type->getNative(), name));
}

Native::VariableReflection* Native::ShaderReflection::findVarByNameInType(TypeReflection* type, const char* name)
{
    return new Native::VariableReflection(m_native->findVarByNameInType(type->getNative(), name));
}

Native::TypeLayoutReflection* Native::ShaderReflection::getTypeLayout(TypeReflection* type, LayoutRules rules)
{
    return new Native::TypeLayoutReflection(m_native->getTypeLayout(type->getNative(), (slang::LayoutRules)rules));
}

Native::TypeReflection* Native::ShaderReflection::specializeType(
    TypeReflection* type,
    SlangInt specializationArgCount,
    TypeReflection* const* specializationArgs,
    ISlangBlob** outDiagnostics)
{

    // Convert Native::TypeReflection array to slang::TypeReflection array
    slang::TypeReflection** nativeArgs = new slang::TypeReflection*[specializationArgCount];
    for (SlangInt i = 0; i < specializationArgCount; i++)
    {
        nativeArgs[i] = (slang::TypeReflection*)specializationArgs[i]->getNative();
    }

    auto callResult = new Native::TypeReflection(m_native->specializeType(
        (slang::TypeReflection*)type->getNative(),
        specializationArgCount,
        nativeArgs,
        outDiagnostics));

    delete[] nativeArgs;
    return callResult;
}

bool Native::ShaderReflection::isSubType(TypeReflection* subType, TypeReflection* superType)
{
    return m_native->isSubType(
        (slang::TypeReflection*)subType->getNative(),
        (slang::TypeReflection*)superType->getNative());
}

SlangUInt Native::ShaderReflection::getHashedStringCount()
{
    return m_native->getHashedStringCount();
}

const char* Native::ShaderReflection::getHashedString(SlangUInt index, size_t* outCount)
{
    return m_native->getHashedString(index, outCount);
}

Native::TypeLayoutReflection* Native::ShaderReflection::getGlobalParamsTypeLayout()
{
    return new Native::TypeLayoutReflection(m_native->getGlobalParamsTypeLayout());
}

Native::VariableLayoutReflection* Native::ShaderReflection::getGlobalParamsVarLayout()
{
    return new Native::VariableLayoutReflection(m_native->getGlobalParamsVarLayout());
}

SlangResult Native::ShaderReflection::toJson(const char** outBlob)
{
    Slang::ComPtr<ISlangBlob> json_blob;
    SlangResult result = m_native->toJson(json_blob.writeRef());
    if (SLANG_FAILED(result))
        throw std::runtime_error("Failed to convert ShaderReflection to JSON");

    // TODO: verify that there is no memory leak here, because it looks like there is
    *outBlob = _strdup((const char*)json_blob->getBufferPointer());

    return SLANG_OK;
}

File: src/Slang.Sdk/Layer2-Binding/Reflection-API/ShaderReflection.cs
using Slang.Sdk.Interop;

namespace Slang.Sdk.Binding
{
    internal unsafe class ShaderReflection : Reflection, IEquatable<ShaderReflection>
    {
        internal Program Program { get; }

        internal override Reflection? Parent => null;
        internal override ShaderReflectionHandle Handle { get; }
        internal override ShaderReflectionHandle NativeHandle => new(StrongInterop.ShaderReflection.GetNative(Handle, out var _));


        internal ShaderReflection(Program program, ShaderReflectionHandle handle)
        {
            Program = program;
            Handle = handle;
        }

        internal uint GetParameterCount()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetParameterCount(Handle, out error), () => error);
        }

        internal uint GetTypeParameterCount()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetTypeParameterCount(Handle, out error), () => error);
        }

        internal TypeParameterReflection GetTypeParameterByIndex(uint index)
        {
            string? error = null;
            return new TypeParameterReflection(this, Call(() => StrongInterop.ShaderReflection.GetTypeParameterByIndex(Handle, index, out error), () => error));
        }

        internal TypeParameterReflection FindTypeParameter(string name)
        {
            string? error = null;
            return new TypeParameterReflection(
                this,
                Call(() => StrongInterop.ShaderReflection.FindTypeParameter(Handle, name, out error), () => error));
        }

        internal VariableLayoutReflection GetParameterByIndex(uint index)
        {
            string? error = null;
            return new VariableLayoutReflection(this, Call(() => StrongInterop.ShaderReflection.GetParameterByIndex(Handle, index, out error), () => error));
        }

        internal uint GetEntryPointCount()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetEntryPointCount(Handle, out error), () => error);
        }

        internal EntryPointReflection GetEntryPointByIndex(uint index)
        {
            string? error = null;
            return new EntryPointReflection(this, Call(() => StrongInterop.ShaderReflection.GetEntryPointByIndex(Handle, index, out error), () => error));
        }

        internal EntryPointReflection FindEntryPointByName(string name)
        {
            string? error = null;
            return new EntryPointReflection(
                this,
                Call(() => StrongInterop.ShaderReflection.FindEntryPointByName(Handle, name, out error), () => error));
        }

        internal uint GetGlobalConstantBufferBinding()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetGlobalConstantBufferBinding(Handle, out error), () => error);
        }

        internal nuint GetGlobalConstantBufferSize()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetGlobalConstantBufferSize(Handle, out error), () => error);
        }

        internal TypeReflection FindTypeByName(string name)
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongInterop.ShaderReflection.FindTypeByName(Handle, name, out error), () => error));
        }

        internal FunctionReflection FindFunctionByName(string name)
        {
            string? error = null;
            return new FunctionReflection(
                this,
                Call(() => StrongInterop.ShaderReflection.FindFunctionByName(Handle, name, out error), () => error));
        }

        internal FunctionReflection FindFunctionByNameInType(TypeReflection type, string name)
        {
            string? error = null;
            return new FunctionReflection(
                this,
                Call(() => StrongInterop.ShaderReflection.FindFunctionByNameInType(Handle, type.Handle, name, out error), () => error));
        }

        internal VariableReflection FindVarByNameInType(TypeReflection type, string name)
        {
            string? error = null;
            return new VariableReflection(
                this,
                Call(() => StrongInterop.ShaderReflection.FindVarByNameInType(Handle, type.Handle, name, out error), () => error));
        }

        internal TypeLayoutReflection GetTypeLayout(TypeReflection type, int layoutRules)
        {
            string? error = null;
            return new TypeLayoutReflection(this, Call(() => StrongInterop.ShaderReflection.GetTypeLayout(Handle, type.Handle, layoutRules, out error), () => error));
        }

        internal TypeReflection SpecializeType(TypeReflection type, int argCount, void** args)
        {
            string? error = null;
            return new TypeReflection(this, Call(() => StrongInterop.ShaderReflection.SpecializeType(Handle, type.Handle, argCount, args, out error), () => error));
        }

        internal bool IsSubType(TypeReflection subType, TypeReflection superType)
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.IsSubType(Handle, subType.Handle, superType.Handle, out error), () => error);
        }

        internal uint GetHashedStringCount()
        {
            string? error = null;
            return Call(() => StrongInterop.ShaderReflection.GetHashedStringCount(Handle, out error), () => error);
        }

        internal string? GetHashedString(uint index)
        {
            string? error = null;
            // Test string?
            return Call(() => StrongInterop.ShaderReflection.GetHashedString(Handle, index, out error), () => error);
        }

        internal TypeLayoutReflection GetGlobalParamsTypeLayout()
        {
            string? error = null;
            return new TypeLayoutReflection(this, Call(() => StrongInterop.ShaderReflection.GetGlobalParamsTypeLayout(Handle, out error), () => error));
        }

        internal VariableLayoutReflection GetGlobalParamsVarLayout()
        {
            string? error = null;
            return new VariableLayoutReflection(this, Call(() => StrongInterop.ShaderReflection.GetGlobalParamsVarLayout(Handle, out error), () => error));
        }

        internal string? ToJson()
        {
            string? error = null;
            return Call(() =>
            {
                StrongInterop.ShaderReflection.ToJson(Handle, out string output, out error);
                return output;
            }, () => error);
        }

        public bool Equals(ShaderReflection? other)
        {
            return this == other;
        }
    }
}
